packages/db/postgres/migrations/004-add-upsert-pins-function.sql

```diff
@@ -0,0 +1,59 @@
+DROP FUNCTION IF EXISTS upsert_pin;
+DROP FUNCTION IF EXISTS upsert_pins;
+
+CREATE OR REPLACE FUNCTION upsert_pin(data json) RETURNS TEXT
+    LANGUAGE plpgsql
+    volatile
+    PARALLEL UNSAFE
+AS
+$$
+DECLARE
+  pin_location_result_id BIGINT;
+  pin_result_id BIGINT;
+BEGIN
+  -- DATA => content_cid, pin(status, location(peer_id, peer_name, region))
+
+  -- Add to pin_location table if new
+  insert into pin_location (peer_id, peer_name, ipfs_peer_id, region)
+  values (data -> 'pin' -> 'location' ->> 'peer_id',
+          data -> 'pin' -> 'location' ->> 'peer_name',
+          data -> 'pin' -> 'location' ->> 'ipfs_peer_id',
+          data -> 'pin' -> 'location' ->> 'region')
+  ON CONFLICT ( peer_id ) DO UPDATE
+          SET "peer_name" = data -> 'pin' -> 'location' ->> 'peer_name',
+              "region" = data -> 'pin' -> 'location' ->> 'region'
+  returning id into pin_location_result_id;
```

Comment:
Let's do the same as @alanshaw did in https://github.com/web3-storage/web3.storage/pull/1101

Reply:
Can do. Why is it `SELECT 42` ?

```
 -- Add to pin_location table if new
  INSERT INTO pin_location (peer_id, peer_name, ipfs_peer_id, region)
    SELECT * FROM (
      SELECT data -> 'pin' -> 'location' ->> 'peer_id' AS peer_id,
             data -> 'pin' -> 'location' ->> 'peer_name' AS peer_name,
             data -> 'pin' -> 'location' ->> 'ipfs_peer_id' AS ipfs_peer_id,
             data -> 'pin' -> 'location' ->> 'region' AS region
    ) AS tmp
    WHERE NOT EXISTS (
      SELECT 42 FROM pin_location WHERE peer_id = data -> 'pin' -> 'location' ->> 'peer_id'
    );
```

Reply:
Ok now doing the same.

Reply:
We talked about this sync.